File: old/Probleme089.py
#! /usr/bin/env python
#-*- coding: utf-8 -*-

import math
import sys
import string

valeur = {'I' : 1, 'V': 5, 'X' : 10, 'L' : 50, 'C' : 100, 'D' : 500, 'M':1000} 

unite = {0: '',1: 'I',2: 'II',3: 'III',4: 'IV',5: 'V',6: 'VI',7: 'VII',8: 'VIII',9: 'IX'}
dizaine = {0: '', 1: 'X', 2: 'XX',3: 'XXX',4: 'XL',5: 'L',6: 'LX',7: 'LXX',8: 'LXXX',9: 'XC'}
centaine = {0: '', 1: 'C',2: 'CC',3: 'CCC',4: 'CD',5: 'D',6: 'DC',7: 'DCC',8: 'DCCC',9: 'CM'}
millier = {0: '',1: 'M',2: 'MM',3: 'MMM',4: 'MMMM',5: 'MMMMM',
	6: 'MMMMMM',7: 'MMMMMMM',8: 'MMMMMMMM',9: 'MMMMMMMMM'}

def convertir(s):
	result = 0
	n = len(s)
	for i in range(n):
		r = valeur[s[i]]
		found = False
		for c in s[(i+1):]:
			if valeur[c] > r:
				found = True
				break
		if not found:
			result = result + r
		else:
			result = result - r
	return result

def romain(n):
	result = ''
	result = result + millier[n/1000]
	n = n%1000
	result = result + centaine[n/100]
	n = n%100
	result = result + dizaine[n/10]
	n = n%10
	result = result + unite[n]
	return result
		

fichier = sys.argv[1]

result = 0
for ligne in open(fichier):
	sold = ligne.strip()
	n = convertir(sold)
	s = romain(n)
	print sold, '->', n, '->', s
	result = result + len(sold) - len(s)

print result

File: problemes/probleme0xx/probleme089.cpp
#include "problemes.h"
#include "arithmetiques.h"

#include <fstream>

typedef long long nombre;
typedef std::vector<nombre> vecteur;

ENREGISTRER_PROBLEME(89, "Roman numerals") {
    // For a number written in Roman numerals to be considered valid there are basic rules which must be followed. Even
    // though the rules allow some numbers to be expressed in more than one way there is always a "best" way of writing
    // a particular number.
    //
    // For example, it would appear that there are at least six ways of writing the number sixteen:
    // 
    //      IIIIIIIIIIIIIIII
    //      VIIIIIIIIIII
    //      VVIIIIII
    //      XIIIIII
    //      VVVI
    //      XVI
    //
    // However, according to the rules only XIIIIII and XVI are valid, and the last example is considered to be the most
    // efficient, as it uses the least number of numerals.
    // 
    // The 11K text file, roman.txt (right click and 'Save Link/Target As...'), contains one thousand numbers written in
    // valid, but not necessarily minimal, Roman numerals; see About... Roman Numerals for the definitive rules for this
    // problem.
    // 
    // Find the number of characters saved by writing each of these in their minimal form.
    // 
    // Note: You can assume that all the Roman numerals in the file contain no more than four consecutive identical
    // units.
    std::map<char, nombre> valeur = {{'I', 1},
                                     {'V', 5},
                                     {'X', 10},
                                     {'L', 50},
                                     {'C', 100},
                                     {'D', 500},
                                     {'M', 1000}};
    const std::vector<std::string> unite = {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};
    const std::vector<std::string> dizaine = {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"};
    const std::vector<std::string> centaine = {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
    const std::vector<std::string> millier = {"", "M", "MM", "MMM", "MMMM", "MMMMM", "MMMMMM", "MMMMMMM", "MMMMMMMM",
                                              "MMMMMMMMM"};

    std::ifstream ifs("data/p089_roman.txt");
    nombre resultat = 0;
    std::string ligne;
    while (ifs >> ligne) {
        nombre n = 0;
        for (auto it = ligne.cbegin(), en = ligne.cend(); it != en; ++it) {
            nombre r = valeur[*it];
            if (std::find_if(it + 1, en, [&r, &valeur](char c) { return valeur[c] > r; }) == en)
                n += r;
            else
                n -= r;
        }

        std::ostringstream oss;
        oss << millier[static_cast<size_t>(n / 1000)];
        n = n % 1000;
        oss << centaine[static_cast<size_t>(n / 100)];
        n = n % 100;
        oss << dizaine[static_cast<size_t>(n / 10)];
        n = n % 10;
        oss << unite[static_cast<size_t>(n)];

        resultat += ligne.size() - oss.str().size();
    }

    return std::to_string(resultat);
}
